src/base/memory_monitor_sr.hpp

```diff
@@ -77,13 +106,111 @@ namespace cubmem
       // and some C++ allocations occurred at header files.
       const int m_magic_number;
   };
+
+  extern memory_monitor *mmon_Gl;
+
+  inline int memory_monitor::get_target_pos ()
+  {
+    return m_target_pos;
+  }
+
+  inline void memory_monitor::make_stat_name (char *buf, const char *file, const int line)
+  {
+    sprintf (buf, "%s:%d", file + m_target_pos, line);
+  }
+
+  inline char *memory_monitor::get_metainfo_pos (char *ptr, size_t size)
+  {
+    return ptr + size - MMON_METAINFO_SIZE;
+  }
+
+  inline void memory_monitor::add_stat (char *ptr, const size_t size, const char *file, const int line)
+  {
+    char stat_name[MMON_MAX_NAME_LENGTH];
+
+    // size should not be 0 because of MMON_METAINFO_SIZE
+    assert (size > 0);
+
+    char *meta_ptr = get_metainfo_pos (ptr, size);
+    MMON_METAINFO *metainfo = (MMON_METAINFO *) meta_ptr;
+
+    metainfo->allocated_size = (uint64_t) size;
+    m_total_mem_usage += metainfo->allocated_size;
+
+    make_stat_name (stat_name, file, line);
+
+retry:
+    const auto search = m_stat_name_map.find (stat_name);
+    if (search != m_stat_name_map.end ())
+      {
+	metainfo->stat_id = search->second;
+	m_stat_map[metainfo->stat_id] += metainfo->allocated_size;
+      }
+    else
+      {
+	std::pair<tbb::concurrent_unordered_map<std::string, int>::iterator, bool> insert_success;
+	metainfo->stat_id = m_stat_name_map.size ();
+	// stat_id starts with 0
+	std::pair <std::string, int> entry (stat_name, metainfo->stat_id);
+	insert_success = m_stat_name_map.insert (entry);
+	if (!insert_success.second)
+	  {
+	    goto retry;
+	  }
+	m_stat_map[metainfo->stat_id] += metainfo->allocated_size;
+      }
+
+    // put meta info into the allocated chunk
+    metainfo->magic_number = m_magic_number;
+    m_meta_alloc_count++;
+  }
+
+  inline void memory_monitor::sub_stat (char *ptr)
+  {
+    size_t allocated_size = malloc_usable_size ((void *)ptr);
+
+    assert (ptr != NULL);
+
+    if (allocated_size >= MMON_METAINFO_SIZE)
+      {
+	char *meta_ptr = get_metainfo_pos (ptr, allocated_size);
+	MMON_METAINFO *metainfo = (MMON_METAINFO *) meta_ptr;
+
+	if (metainfo->magic_number == m_magic_number)
+	  {
+	    assert ((metainfo->stat_id >= 0 && metainfo->stat_id < m_stat_map.size()));
+	    assert (m_stat_map[metainfo->stat_id] >= metainfo->allocated_size);
+	    assert (m_total_mem_usage >= metainfo->allocated_size);
```

Comment:
It would be better to combine the assert statements performed using metainfo->allocated_size into one using && conditions.

```suggestion
	    assert (m_total_mem_usage >= metainfo->allocated_size);
	    assert (metainfo->allocated_size == allocated_size);
```

Reply:
Done.